File: sortedunknown.java
// Time Complexity : O(log n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

public class sortedunknown {
    
    public int search(ArrayReader reader, int target){

        int low = 0, high = 1;

        while(reader.get(high) <= target){
            low = high;
            high = high * 2;

        }
        return binarySearch(reader, target, low, high);
    }

    private int binarySearch(ArrayReader reader, int target, int low, int high){
        while(low <= high){
            int mid = low + (high - low) / 2;

            //base case if the middle is the target
            if(reader.get(mid) == target){
                return mid;
            }
            else if(reader.get(mid) > target){
                //move left
                high = mid - 1;
            }
            else{
                //move to the right
                low = mid + 1;
            }
            return -1;
        }
    }
}

File: sortedunknown.py
##Time Complexity : O(n)
##Space Complexity : O(1)
##Did this code successfully run on Leetcode : Yes
class Solution:
    def search(self, nums, target):
        left = 0
        right = len(nums) - 1
        while left <= right:
            middle = (left + right) // 2
            if nums[middle] == target:
                return middle

            if nums[left] <= nums[middle]:
                if nums[left] <= target <= nums[middle]:
                    right = middle - 1
                else:
                    left = middle + 1  
            else:
                if nums[middle] <= target <= nums[right]:
                    left = middle + 1 
                else:
                    right = middle - 1  
        return -1